### PR TITLE
Chore(filter v2) test updates

### PR DESCRIPTION
**Pull request by romanzac**
Description
Batch of new tests for Waku Filter v2.

Changes
Tests available to run:
cd go-waku/waku/v2/protocol/filter
go test

Two tests failing at the moment - corresponding issues were opened.

### CODE REVIEW

File: waku/v2/protocol/filter/filter_test.go [hunk 1]
```diff
@@ -170,6 +271,16 @@ func (s *FilterTestSuite) publishMsg(topic, contentTopic string, optionalPayload
 	s.Require().NoError(err)
 }
 
+func (s *FilterTestSuite) publishMessages(msgs []WakuMsg) {
+	for _, m := range msgs {
+		_, err := s.relayNode.PublishToTopic(s.ctx, tests.CreateWakuMessage(m.contentTopic, utils.GetUnixEpoch(), m.payload), m.pubSubTopic)
+		s.Require().NoError(err)
+		//dst := make([]byte, hex.EncodedLen(len(hash)))
+		//hex.Encode(dst, hash)
+		//s.log.Info("Published message ", zap.String("hash:", string(dst)))
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:280]:
```suggestion
```

File: waku/v2/protocol/filter/filter_test.go [hunk 2]
```diff
@@ -125,6 +142,60 @@ func (s *FilterTestSuite) waitForMsg(fn func(), ch chan *protocol.Envelope) {
...
+				select {
+				case env := <-sub.C:
+					received := WakuMsg{
+						pubSubTopic:  env.PubsubTopic(),
+						contentTopic: env.Message().GetContentTopic(),
+						payload:      string(env.Message().GetPayload()),
+					}
+					s.log.Info("Received message ", zap.String("pubSubTopic", received.pubSubTopic))
+					s.log.Info("Received message ", zap.String("contentTopic", received.contentTopic))
+					s.log.Info("Received message ", zap.String("payload", received.payload))
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:178]:
```suggestion
					s.log.Info("received message ", zap.String("pubSubTopic", received.pubSubTopic),  zap.String("contentTopic", received.contentTopic), zap.String("payload", received.payload))
```

File: waku/v2/protocol/filter/filter_test.go [hunk 3]
```diff
@@ -520,6 +645,408 @@ func (s *FilterTestSuite) TestAutoShard() {
...
+	s.ctxCancel = cancel
+
+	// Prepare data
+	for i := 0; i < 3; i++ {
+		suffix = fmt.Sprintf("%02d", i)
+		messages = append(messages, WakuMsg{
+			pubSubTopic:  testTopic,
+			contentTopic: testContentTopic + suffix,
+			payload:      testPayload,
+		})
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:777]:
```suggestion
		messages = append(messages, WakuMsg{
			pubSubTopic:  testTopic,
			contentTopic: fmt.Sprintf("%s%02d", testContentTopic, i),
			payload:      testPayload,
		})
```

File: waku/v2/protocol/filter/filter_test.go [hunk 4]
```diff
@@ -520,6 +645,408 @@ func (s *FilterTestSuite) TestAutoShard() {
...
+	var (
+		testTopic        = "/waku/2/go/filter/test"
+		testContentTopic = "Topic"
+		testPayload      = "test_msg"
+		suffix           string
+		messages         []WakuMsg
+	)
+
+	// Create test context
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second) // Test can't exceed 10 seconds
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:805]:
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second) // Test can't exceed 20 seconds
```

File: waku/v2/protocol/filter/filter_test.go [hunk 5]
```diff
@@ -520,6 +645,408 @@ func (s *FilterTestSuite) TestAutoShard() {
...
+		s.subDetails = append(s.subDetails, s.subscribe(m.pubSubTopic, m.contentTopic, s.fullNodeHost.ID())...)
+		s.log.Info("Subscribing ", zap.String("PubSubTopic", m.pubSubTopic))
+		_, err := s.relayNode.SubscribeToTopic(context.Background(), m.pubSubTopic)
+		s.Require().NoError(err)
+	}
+
+	// Debug to see subscriptions in light node
+	for _, sub := range s.subDetails {
+		s.log.Info("Light Node subscription ", zap.String("PubSubTopic", sub.ContentFilter.PubsubTopic))
+		s.log.Info("Light Node subscription ", zap.String("ContentTopic", sub.ContentFilter.ContentTopicsList()[0]))
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:830]:
```suggestion
		s.log.Info("Light Node subscription ", zap.String("PubSubTopic", sub.ContentFilter.PubsubTopic), zap.String("ContentTopic", sub.ContentFilter.ContentTopicsList()[0]))
```

File: waku/v2/protocol/filter/filter_test.go [hunk 6]
```diff
@@ -520,6 +645,408 @@ func (s *FilterTestSuite) TestAutoShard() {
...
+	var (
+		testTopic        = "/waku/2/go/filter/test"
+		testContentTopic = "Topic"
+		testPayload      = "test_msg"
+		suffix           string
+		messages         []WakuMsg
+	)
+
+	// Create test context
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second) // Test can't exceed 10 seconds
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:853]:
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second) // Test can't exceed 20 seconds
```

File: waku/v2/protocol/filter/filter_test.go [hunk 7]
```diff
@@ -520,6 +645,408 @@ func (s *FilterTestSuite) TestAutoShard() {
...
+	}
+
+	// Prepare data
+	for i := 0; i < maxContentTopics; i++ {
+		suffix = fmt.Sprintf("%02d", i)
+		messages = append(messages, WakuMsg{
+			pubSubTopic:  testTopic,
+			contentTopic: testContentTopic + suffix,
+			payload:      testPayload + suffix,
+		})
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:945]: Actually, this code seems to repeat very frequently? perhaps it's good to extract it into a non exported function

File: waku/v2/protocol/filter/filter_test.go [hunk 8]
```diff
@@ -520,6 +645,408 @@ func (s *FilterTestSuite) TestAutoShard() {
...
+	s.ctx = ctx
+	s.ctxCancel = cancel
+
+	// Prepare data
+	for i := 0; i < 10; i++ {
+		suffix = fmt.Sprintf("%02d", i)
+		messages = append(messages, WakuMsg{
+			pubSubTopic:  testTopic,
+			contentTopic: testContentTopic + suffix,
+			payload:      testPayload + suffix,
```
Comment (richard-ramos) [inline waku/v2/protocol/filter/filter_test.go:863]:
```suggestion
		suffix = fmt.Sprintf("%02d", i)
		messages = append(messages, WakuMsg{
			pubSubTopic:  testTopic,
			contentTopic:  fmt.Sprintf("%s%02d", testContentTopic, i),
			payload:       fmt.Sprintf("%s%02d", testPayload, i),
```

**Review by richard-ramos — COMMENTED** [review]
Left some nits/suggestions

File: waku/v2/protocol/filter/filter_test.go (hunk 7 above)
Comment (romanzac) [reply waku/v2/protocol/filter/filter_test.go]: Refactored in b5715d6a29aeef36a5b29c0db4caf20c7fdcd93d

**Comment by chaitanyaprem** [comment]
@romanzac , Was going through the failed test `TestIncorrectSubscribeIdentifier` and had noticed that there is no reason for this test to fail. As the filterLightNode opens a stream with the correct subscribe protocol as and when we invoke the subscribe API. 
https://github.com/waku-org/go-waku/blob/9161c4f7fee3ff5971489547f0d618ae6c158520/waku/v2/protocol/filter/client.go#L169

So even if we add a wrong protocol to the client, that would just be an additional protocol supported by this lightClient and does not affect how it communicates with filter node. 
If you want to simulate a test where a wrong protocolID is used for subscribe, then you would have to initiate FilterClient itself using libp2p API directly and then form a subscribe message and send it over a stream opened with the incorrect protocolID.

**Comment by chaitanyaprem** [comment]
@richard-ramos , for some reason Test for WakuMetadata proto seems to also fail. 
Could it be running into some sort of special condition where the test fails? Would be better to rootcause and fix it.

**Comment by richard-ramos** [comment]
> @richard-ramos , for some reason Test for WakuMetadata proto seems to also fail.
> Could it be running into some sort of special condition where the test fails? Would be better to rootcause and fix it.

Fixed in https://github.com/waku-org/go-waku/pull/826

**Comment by romanzac** [comment]
@chaitanyaprem Filter tests are passing on my side. I have followed your suggestions for "TestIncorrectSubscribeIdentifier". It is working, although I had to go deep into the stack and copied a lot of code. Not sure, if by chance you would find beneficial to have subscribe protocol identifier changeable at higher levels. Or perhaps Peerstore().RemoveProtocols() might work ?

**Comment by chaitanyaprem** [comment]
> @chaitanyaprem Filter tests are passing on my side. I have followed your suggestions for "TestIncorrectSubscribeIdentifier". It is working, although I had to go deep into the stack and copied a lot of code. Not sure, if by chance you would find beneficial to have subscribe protocol identifier changeable at higher levels. Or perhaps Peerstore().RemoveProtocols() might work ?

Actually RemoveProtocols might have worked, slipped my mind. Also, do note that this is a kind of test you are trying to do which may sometimes require emulating lot of client code(as you are trying to simulate a scenario which would not be possible with existing client implementations)

**Comment by chaitanyaprem** [comment]
Also looks like #836 seems to have slipped past your tests and mine.
Might be a good idea to check the scope to see such scenarios are covered.

Haven't reviewed all your tests though, so not sure if it is already covered. :)

**Comment by romanzac** [comment]
> Also looks like #836 seems to have slipped past your tests and mine. Might be a good idea to check the scope to see such scenarios are covered.
> 
> Haven't reviewed all your tests though, so not sure if it is already covered. :)

Do you mean in general, something is called with initialized variable, but no specific values set ? I will come to these later.

**Comment by romanzac** [comment]
> > @chaitanyaprem Filter tests are passing on my side. I have followed your suggestions for "TestIncorrectSubscribeIdentifier". It is working, although I had to go deep into the stack and copied a lot of code. Not sure, if by chance you would find beneficial to have subscribe protocol identifier changeable at higher levels. Or perhaps Peerstore().RemoveProtocols() might work ?
> 
> Actually RemoveProtocols might have worked, slipped my mind. Also, do note that this is a kind of test you are trying to do which may sometimes require emulating lot of client code(as you are trying to simulate a scenario which would not be possible with existing client implementations)

I will reuse those "copied" funcs in later tests where I go down to the level of invalid payloads. 

BTW: I have tried RemoveProtocols and it didn't work for me. I see from the code subscribe related protocol identifier is not part of a live node instance. It is part of the code only. Could you confirm for me please ?

**Comment by romanzac** [comment]
Guys, I wish we can close this PR by Friday this week. Its scope is too large. I will create smaller PRs starting next week.

**Comment by chaitanyaprem** [comment]
> > > @chaitanyaprem Filter tests are passing on my side. I have followed your suggestions for "TestIncorrectSubscribeIdentifier". It is working, although I had to go deep into the stack and copied a lot of code. Not sure, if by chance you would find beneficial to have subscribe protocol identifier changeable at higher levels. Or perhaps Peerstore().RemoveProtocols() might work ?
> > 
> > 
> > Actually RemoveProtocols might have worked, slipped my mind. Also, do note that this is a kind of test you are trying to do which may sometimes require emulating lot of client code(as you are trying to simulate a scenario which would not be possible with existing client implementations)
> 
> I will reuse those "copied" funcs in later tests where I go down to the level of invalid payloads.
> 
> BTW: I have tried RemoveProtocols and it didn't work for me. I see from the code subscribe related protocol identifier is not part of a live node instance. It is part of the code only. Could you confirm for me please ?

Oh right, since the protocol stream is created only when a request has to be sent which means anytime call to subscribe API will initiate a new stream with the protocol and hence not possible.

**Comment by romanzac** [comment]
May I proceed with merge to master ?

**Comment by richard-ramos** [comment]
:( We shouldn't merge PRs that have not been approved

**Comment by romanzac** [comment]
> :( We shouldn't merge PRs that have not been approved

I am sorry, I've interpreted silence for one day as yes. I've announced I would like to close it by Friday and approval was not hardwired as precondition to merge. I will ping you on Discord next time.

**Pull request merged**